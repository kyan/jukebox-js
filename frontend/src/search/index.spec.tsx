import React from 'react'
import { Provider } from 'react-redux'
import { mount } from 'enzyme'
import configureMockStore from 'redux-mock-store'
import SearchContainer from './index'

describe('SearchContainer', () => {
  const mockStore = configureMockStore()

  beforeEach(() => {
    jest.clearAllMocks()
  })

  describe('render', () => {
    test('Adding a track works as expected', () => {
      const store = mockStore({
        search: {
          activePage: 2,
          limit: 20,
          query: 'happy',
          total: 3200,
          results: [
            {
              track: {
                name: 'Track name 1',
                uri: 'https://open.spotify.com/track/0c41pMosF5Kqwwetrack1',
                artist: {
                  name: 'Artist name 1'
                },
                album: {
                  name: 'Album name 1'
                },
                image: 'image1'
              }
            }
          ],
          searchSideBarOpen: true
        },
        curatedList: {
          tracks: []
        }
      })
      const wrapper = mount(
        <Provider store={store}>
          <SearchContainer />
        </Provider>
      )
      const actions = store.getActions()
      const event = { target: { value: 'spam' } }
      wrapper.find('SidebarPusher').simulate('click')
      wrapper.find('Form').simulate('submit')
      wrapper.find('.search-list-item .search-list-item__image').first().simulate('click')
      wrapper.find('input').first().simulate('change', event)
      wrapper.find('Pagination').find('.item').last().simulate('click')
      expect(actions).toMatchSnapshot()
    })

    test('Adding a track to the mix works as expected', () => {
      const store = mockStore({
        search: {
          activePage: 2,
          limit: 20,
          query: 'happy',
          total: 3200,
          results: [
            {
              track: {
                name: 'Track name 1',
                uri: 'https://open.spotify.com/track/0c41pMosF5Kqwwetrack1',
                artist: {
                  name: 'Artist name 1'
                },
                album: {
                  name: 'Album name 1'
                },
                image: 'image1'
              }
            }
          ],
          searchSideBarOpen: true
        },
        curatedList: {
          tracks: []
        }
      })
      const wrapper = mount(
        <Provider store={store}>
          <SearchContainer />
        </Provider>
      )
      const actions = store.getActions()
      wrapper.find('SidebarPusher').simulate('click')
      wrapper.find('.search-list-item .search-list-item__add').first().simulate('click')
      expect(actions).toMatchSnapshot()
    })

    test('Removing a track from the mix works as expected', () => {
      const store = mockStore({
        search: {
          activePage: 2,
          limit: 20,
          query: 'happy',
          total: 3200,
          results: [
            {
              track: {
                name: 'Track name 1',
                uri: 'https://open.spotify.com/track/0c41pMosF5Kqwwetrack1',
                artist: {
                  name: 'Artist name 1'
                },
                album: {
                  name: 'Album name 1'
                },
                image: 'image1'
              }
            }
          ],
          searchSideBarOpen: true
        },
        curatedList: {
          tracks: [
            {
              track: {
                name: 'Track name 2',
                uri: 'https://open.spotify.com/track/0c41pMosF5Kqwwetrack2',
                artist: {
                  name: 'Artist name 2'
                },
                album: {
                  name: 'Album name 2'
                },
                image: 'image2'
              }
            }
          ]
        }
      })
      const wrapper = mount(
        <Provider store={store}>
          <SearchContainer />
        </Provider>
      )
      const actions = store.getActions()
      wrapper.find('SidebarPusher').simulate('click')
      wrapper
        .find('.search-list-item-draggable .search-list-item__remove')
        .first()
        .simulate('click')
      expect(actions).toMatchSnapshot()
    })

    test('Adding tracks from the mix works as expected', () => {
      const store = mockStore({
        search: {
          activePage: 2,
          limit: 20,
          query: 'happy',
          total: 3200,
          results: [
            {
              track: {
                name: 'Track name 1',
                uri: 'https://open.spotify.com/track/0c41pMosF5Kqwwetrack1',
                artist: {
                  name: 'Artist name 1'
                },
                album: {
                  name: 'Album name 1'
                },
                image: 'image1'
              }
            }
          ],
          searchSideBarOpen: true
        },
        curatedList: {
          tracks: [
            {
              track: {
                name: 'Track name 2',
                uri: 'https://open.spotify.com/track/0c41pMosF5Kqwwetrack2',
                artist: {
                  name: 'Artist name 2'
                },
                album: {
                  name: 'Album name 2'
                },
                image: 'image2'
              }
            },
            {
              track: {
                name: 'Track name 3',
                uri: 'https://open.spotify.com/track/0c41pMosF5Kqwwetrack3',
                artist: {
                  name: 'Artist name 3'
                },
                album: {
                  name: 'Album name 3'
                },
                image: 'image3'
              }
            }
          ]
        }
      })
      const wrapper = mount(
        <Provider store={store}>
          <SearchContainer />
        </Provider>
      )
      const actions = store.getActions()
      wrapper.find('SidebarPusher').simulate('click')
      wrapper.find('.search-list-item__add_to_mix').first().simulate('click')
      expect(actions).toMatchSnapshot()
    })

    test('Swapping tracks from the mix works as expected', () => {
      const store = mockStore({
        search: {
          activePage: 2,
          limit: 20,
          query: 'happy',
          total: 3200,
          results: [],
          searchSideBarOpen: true
        },
        curatedList: {
          tracks: [
            {
              track: {
                name: 'Track name 3',
                uri: 'https://open.spotify.com/track/0c41pMosF5Kqwwetrack3',
                artist: {
                  name: 'Artist name 3'
                },
                album: {
                  name: 'Album name 3'
                },
                image: 'image3'
              }
            }
          ]
        }
      })
      const wrapper = mount(
        <Provider store={store}>
          <SearchContainer />
        </Provider>
      )
      const actions = store.getActions()
      //@ts-ignore
      wrapper.find('DraggableSearchItem').props().action('1', '3')
      expect(actions).toMatchSnapshot()
    })

    test('Does not add when more than 5 already added', () => {
      const store = mockStore({
        search: {
          activePage: 2,
          limit: 20,
          query: 'happy',
          total: 3200,
          results: [
            {
              track: {
                name: 'Track name 1',
                uri: 'https://open.spotify.com/track/0c41pMosF5Kqwwetrack1',
                artist: {
                  name: 'Artist name 1'
                },
                album: {
                  name: 'Album name 1'
                },
                image: 'image1'
              }
            }
          ],
          searchSideBarOpen: true
        },
        curatedList: {
          tracks: [
            {
              track: {
                name: 'Track name 2',
                uri: 'https://open.spotify.com/track/0c41pMosF5Kqwwetrack2',
                artist: {
                  name: 'Artist name 2'
                },
                album: {
                  name: 'Album name 2'
                },
                image: 'image2'
              }
            },
            {
              track: {
                name: 'Track name 3',
                uri: 'https://open.spotify.com/track/0c41pMosF5Kqwwetrack3',
                artist: {
                  name: 'Artist name 3'
                },
                album: {
                  name: 'Album name 3'
                },
                image: 'image3'
              }
            },
            {
              track: {
                name: 'Track name 4',
                uri: 'https://open.spotify.com/track/0c41pMosF5Kqwwetrack4',
                artist: {
                  name: 'Artist name 4'
                },
                album: {
                  name: 'Album name 4'
                },
                image: 'image4'
              }
            },
            {
              track: {
                name: 'Track name 5',
                uri: 'https://open.spotify.com/track/0c41pMosF5Kqwwetrack5',
                artist: {
                  name: 'Artist name 5'
                },
                album: {
                  name: 'Album name 5'
                },
                image: 'image5'
              }
            },
            {
              track: {
                name: 'Track name 6',
                uri: 'https://open.spotify.com/track/0c41pMosF5Kqwwetrack6',
                artist: {
                  name: 'Artist name 6'
                },
                album: {
                  name: 'Album name 6'
                },
                image: 'image6'
              }
            }
          ]
        }
      })
      const wrapper = mount(
        <Provider store={store}>
          <SearchContainer />
        </Provider>
      )
      const actions = store.getActions()
      wrapper.find('SidebarPusher').simulate('click')
      wrapper.find('.search-list-item .search-list-item__add').first().simulate('click')
      expect(actions).toMatchSnapshot()
    })
  })
})
