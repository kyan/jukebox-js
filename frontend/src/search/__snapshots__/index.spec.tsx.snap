// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchContainer render Adding a track to the mix works as expected 1`] = `
Array [
  Object {
    "type": "actionRemoveFromSearchResults",
    "uris": Array [
      "https://open.spotify.com/track/0c41pMosF5Kqwwetrack1",
    ],
  },
  Object {
    "track": Object {
      "album": Object {
        "name": "Album name 1",
      },
      "artist": Object {
        "name": "Artist name 1",
      },
      "image": "image1",
      "name": "Track name 1",
      "uri": "https://open.spotify.com/track/0c41pMosF5Kqwwetrack1",
    },
    "type": "curateAddTrackToMix",
  },
]
`;

exports[`SearchContainer render Adding a track works as expected 1`] = `
Array [
  Object {
    "open": true,
    "type": "actionToggleSearchSidebar",
  },
  Object {
    "key": "searchGetTracks",
    "params": Object {
      "options": Object {
        "limit": 20,
        "offset": 0,
      },
      "query": "happy",
    },
    "type": "actionTrackSearch",
  },
  Object {
    "type": "actionRemoveFromSearchResults",
    "uris": Array [
      "https://open.spotify.com/track/0c41pMosF5Kqwwetrack1",
    ],
  },
  Object {
    "key": "tracklist.add",
    "params": Object {
      "uris": Array [
        "spotify:track:0c41pMosF5Kqwwetrack1",
      ],
    },
    "type": "actionSend",
  },
  Object {
    "query": "spam",
    "type": "actionStoreSearchQuery",
  },
  Object {
    "key": "searchGetTracks",
    "params": Object {
      "options": Object {
        "limit": 20,
        "offset": 20,
      },
      "query": "happy",
    },
    "type": "actionTrackSearch",
  },
]
`;

exports[`SearchContainer render Adding tracks from the mix works as expected 1`] = `
Array [
  Object {
    "key": "tracklist.add",
    "params": Object {
      "uris": Array [
        "spotify:track:0c41pMosF5Kqwwetrack2",
        "spotify:track:0c41pMosF5Kqwwetrack3",
      ],
    },
    "type": "actionSend",
  },
  Object {
    "type": "searchClearMix",
  },
]
`;

exports[`SearchContainer render Does not add when more than 5 already added 1`] = `
Array [
  Object {
    "open": true,
    "type": "actionToggleSearchSidebar",
  },
  Object {
    "open": false,
    "type": "actionToggleSearchSidebar",
  },
]
`;

exports[`SearchContainer render Removing a track from the mix works as expected 1`] = `
Array [
  Object {
    "type": "curateRemoveTracksFromMix",
    "uri": "https://open.spotify.com/track/0c41pMosF5Kqwwetrack2",
  },
  Object {
    "key": "searchGetTracks",
    "params": Object {
      "options": Object {
        "limit": 20,
        "offset": 0,
      },
      "query": "happy",
    },
    "type": "actionTrackSearch",
  },
]
`;

exports[`SearchContainer render Swapping tracks from the mix works as expected 1`] = `
Array [
  Object {
    "a": 2,
    "b": 1,
    "type": "searchSwapTracks",
  },
]
`;
